### DISCUSSION TITLE
Rotary encoder problems

### DISCUSSION THREAD

**Discussion by Lorca-Squirrel** [DCC-EX/EX-T3-WiFi#89] [Q&A] [answered]
I have the throttle partially completed.  I think all of the code uploaded properly but I am having some problems with the rotary encoder.  When turning it, it will sometimes function normally but usually it will not register every step or it will jump steps.  It will also change the direction of travel a lot (the switch is not being pressed).  Any ideas?

https://user-images.githubusercontent.com/130647217/231704814-554f1313-5351-46e2-8dfe-0e1380da5253.mp4

**Comment by alex-code** [comment]
Could you show your wiring of the encoder to the Firebeetle2?

**Comment by Lorca-Squirrel** [comment]
I followed the wiring scheme in the assembly section.  I also tried using a different rotary encoder and it did the same thing.
![20230414_093106](https://user-images.githubusercontent.com/130647217/232074953-9e125860-900c-4a28-95e4-00de4f896bc4.jpg)
![20230414_093106(0)](https://user-images.githubusercontent.com/130647217/232074968-699d9760-e9f0-4eff-a36c-93cfa16c7e81.jpg)
![20230414_093053(0)](https://user-images.githubusercontent.com/130647217/232074973-3c13c57f-e221-4550-978d-bd35045e7e8f.jpg)

**Reply by alex-code** [reply]
That looks right.
Do the jumps only happen when turning the encoder?
Do you know what version your CS is?

**Reply by Lorca-Squirrel** [reply]
They happen only when turning the encoder and I am using the latest version of DCC++Ex (4.1.5) with the myfilter.cpp file.

**Reply by alex-code** [reply]
Would you be able to connect your computer to the CS and open a serial monitor so we can see the commands that are being received.

**Comment by Lorca-Squirrel** [comment]
https://user-images.githubusercontent.com/130647217/232168767-bf8b59ec-8e33-434d-8c02-fb73a217218f.mp4

**Comment by Lorca-Squirrel** [comment]
Also I did try completely wiping the ESP32 and reinstalling both the Filesystem Image and the code itself.

**Reply by alex-code** [reply]
That's an odd one, is the throttle code you uploaded recent or a bit older?

**Reply by Lorca-Squirrel** [reply]
I downloaded it fresh earlier today when I tried reinstalling it.

**Reply by Lorca-Squirrel** [reply]
It should be the most recent version.  I accessed it from the link in the DCC++Ex project

**Reply by Lorca-Squirrel** [reply]
I just tried disconnecting the wire going to pad c, (the center button press on the rotary encoder) and it will still flip directions randomly  when rotated in addition to jumping.

**Reply by alex-code** [reply]
Do you have a multimeter?
Would you be able to do a continuity test on the wires?
This seems a lot like interference so there could be some crossover.

**Reply by Lorca-Squirrel** [reply]
I can do that.  Between which points?

**Reply by Lorca-Squirrel** [reply]
I have also found a second issue that might be related.  If I set the throttle to a speed step, a couple seconds later it will go back to 0.

**Reply by alex-code** [reply]
> I can do that. Between which points?

I'd probe one end while then probing all the other wires to make sure no two are interfering with eachother.
Do this one by one until all have been probed.

There's nothing else connected to the CS is there? Like JMRI?

**Reply by Lorca-Squirrel** [reply]
I just probed the ESP and pins A1, A2, and A3 seem to be bridged together.  A5 and A0 seem to be unaffected

**Reply by Lorca-Squirrel** [reply]
would it be possible to swap the functions of pin A3 (C on the rotary encoder) to pin A4 since it doesnt appear to be effected.

**Reply by Lorca-Squirrel** [reply]
It looks like the Firebeetle is defective.  I looked at the schematic for it and I dont see anything that would connect A1, A2, and A3 together.

**Reply by alex-code** [reply]
It could be the encoder bridging the pins.
You could try removing the encoder and probing again.

I'd have to check the schematic about the other pins as although they're exposed some are used by the firebeetle2 for other things like the display.

**Reply by Lorca-Squirrel** [reply]
Removed both the encoder and the display and the connection is still there.

https://user-images.githubusercontent.com/130647217/232288218-4bd43e95-f5dc-4a97-8d5d-7276d417fd9c.mp4

**Reply by alex-code** [reply]
That does seem defective, are there any visible bits of solder possibly bridging the pins?
If not I'd say you'll need to exchange it.

**Reply by daveclifford** [reply]
I'd get the one with the header already soldered - less risk of a bridge through soldering.

Dave

**Reply by Lorca-Squirrel** [reply]
Went ahead and contacted DFrobot.  I dont see any solder bridging the pins so I dont think its that but it could even be as simple as some solder paste left under the ESP32 itself.

**Comment by Lorca-Squirrel** [comment, marked as answer]
Just to cap off this discussion.  The ESP32 is in fact defective.  I have a replacement on the way.  Thank you for the assistance!

**Reply by alex-code** [reply]
Thanks for confirming, hopefully you'll have more luck with a new one.
